src/test/java/net/nuagenetworks/bambou/RestCertificateTest.java

```diff
@@ -0,0 +1,244 @@
+        
+        KeyManager[] keyManagers = DynamicKeystoreGenerator.generateKeyManagersForDynamicKeystore(keystore);
+        Assert.assertEquals(1, keyManagers.length);
+        X509ExtendedKeyManager ekm = (X509ExtendedKeyManager) keyManagers[0];
+        
+        String[] aliases = ekm.getClientAliases("RSA", null);
+        String alias = aliases[0];
+        Assert.assertEquals("Keymanager Keystore client alias ", generatorAlias, alias);
+        
+        PrivateKey keyManagerPrivateKey = ekm.getPrivateKey(alias);
+        Assert.assertEquals("Keymanager RSA private key format", "PKCS#8", keyManagerPrivateKey.getFormat());
+        Assert.assertEquals("Keymanager RSA private key algorithm", "RSA", keyManagerPrivateKey.getAlgorithm());
+        Assert.assertEquals("Keymanager RSA private key length", 46, keyManagerPrivateKey.toString().length());
+        
+        X509Certificate[] keyManagerCertificates = ekm.getCertificateChain(generatorAlias);
+        X509Certificate keyManagerCert = keyManagerCertificates[0];
+        Assert.assertEquals("Keymanager certificate chain length", 1, keyManagerCertificates.length);
+        Assert.assertEquals("Keymanager certificate type", "X.509", keyManagerCert.getType());
+        Assert.assertEquals("Keymanager certificate signature algorithm", "SHA256withRSA", keyManagerCert.getSigAlgName());
+    }
+    
+    @Test
+    public void testBadKey() {
+		byte[] decodedPrivateKey = DynamicKeystoreGenerator.getDecodedPrivateKey(BAD_TESTABLE_PRIVATE_KEY);
+		int len = decodedPrivateKey.length;
+
+        Assert.assertEquals("Check private key decoded length", 1123, len);
+        
+        try {
+            RSAPrivateKey rsaPrivateKey = DynamicKeystoreGenerator.generateRSAPrivateKey(decodedPrivateKey);
+            Assert.assertFalse(false);
+        } catch (KeyManagementException ex) {
+            Assert.assertThat(ex.getMessage(), CoreMatchers.containsString("InvalidKeySpecException"));
+            Assert.assertThat(ex.getMessage(), CoreMatchers.containsString("InvalidKeyException"));
+            Assert.assertThat(ex.getMessage(), CoreMatchers.containsString("IOException"));
+            Assert.assertThat(ex.getMessage(), CoreMatchers.containsString("Detect premature EOF"));
+        }
+    }
+    
+    @Test
+    public void testBadCertificate() {
+    	try {
+    		InputStream is = new ByteArrayInputStream(BAD_TESTABLE_CERTIFICATE.getBytes());
+    		DynamicKeystoreGenerator.generateX509Certificate(is);
+    		Assert.assertFalse(false);
+    	} catch (KeyManagementException ex) {
+    		Assert.assertThat(ex.getMessage(), CoreMatchers.containsString("CertificateException"));
+    		Assert.assertThat(ex.getMessage(), CoreMatchers.containsString("IOException"));
+    		Assert.assertThat(ex.getMessage(), CoreMatchers.containsString("extra data given to DerValue constructor"));
+    	}
+    }
+}
```

Comment:
In addition to these unit tests, any chance the good/bad certificate scenarios can be tested through java-bambou's API. The test(s) would verify that the right methods are called on DynamicKeystoreGenerator (with the right arguments/values, in the right sequence), for example? DynamicKeystoreGenenator could be stubbed out and interceptors defined for the methods of interest. Or could it be done at a lower level, by intercepting calls in the javax.security.* interfaces/classes?